android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/ResourceExtension.kt

```diff
@@ -315,3 +325,48 @@ fun String.extractLogicalIdUuid() = this.substringAfter("/").substringBefore("/"
 fun Resource.addTags(tags: List<Coding>) {
   tags.forEach { this.meta.addTag(it) }
 }
+
+suspend fun Task.updateDependentTaskDueDate(defaultRepository: DefaultRepository): Task {
+  return apply {
+    val isUpcoming = status == Task.TaskStatus.REQUESTED
+    val isOverdue =
+      (status == Task.TaskStatus.READY || status == Task.TaskStatus.INPROGRESS) &&
+        executionPeriod.end.before(Date())
+    val isDue = status == Task.TaskStatus.INPROGRESS
+    if (!(isUpcoming || isOverdue || isDue) || partOf.isEmpty()) {
+      return this
+    }
+    partOf.forEach { task ->
+      val dependentTask = defaultRepository.loadResource(task.reference as Reference) as? Task
+      dependentTask?.output?.forEach { dependentTaskOutputValue ->
+        val encounterResource =
+          defaultRepository.loadResource(
+            readValueFromJsonString(dependentTaskOutputValue.valueToString(), "reference")
+          ) as?
+            Encounter
+        encounterResource?.partOf?.reference?.let { partOfReference ->
+          val immunizationResource =
+            defaultRepository.loadResource(partOfReference) as? Immunization
+          immunizationResource?.occurrenceDateTimeType?.dateTimeValue()?.valueAsCalendar?.let {
+            immunizationDate ->
+            val dependentTaskStartDate = dependentTask.executionPeriod.start.toInstant()
+            val difference =
+              Duration.between(immunizationDate.toInstant(), dependentTaskStartDate).toDays()
+            if (difference < dependentTask.input.first().valueToString().toInt()) {
+              dependentTask.executionPeriod.start =
+                Date.from(
+                  LocalDate.parse(immunizationDate.toString())
+                    .plusDays(dependentTask.input.first().valueToString().toLong())
+                    .atStartOfDay()
+                    .toInstant(ZoneOffset.UTC)
```

Comment:
do we have a helper to parse this? we should make one if not to make sure we parse consistently